File: android/src/main/kotlin/ru/kardenvan/flutter_yandex_ads_plus/platform_api/ads_organizer/RewardedYandexAdsOrganizer.kt
package ru.kardenvan.flutter_yandex_ads_plus.platform_api.ads_organizer

import android.content.Context
import ru.kardenvan.flutter_yandex_ads_plus.ads.rewarded_yandex_ad.RewardedYandexAd
import ru.kardenvan.flutter_yandex_ads_plus.ads.rewarded_yandex_ad.RewardedYandexAdArguments
import ru.kardenvan.flutter_yandex_ads_plus.platform_api.ad_event_dispatcher.FlutterYandexAdsEventDispatcher
import ru.kardenvan.flutter_yandex_ads_plus.platform_api.ad_event_dispatcher_facade.RewardedAdEventDispatcherFacade

class RewardedYandexAdsOrganizer(
    private val eventDispatcher: FlutterYandexAdsEventDispatcher,
) {
    private val ads: MutableMap<String, RewardedYandexAd> = mutableMapOf()

    fun createAndLoadAd(
        context: Context,
        arguments: RewardedYandexAdArguments,
    ) {
        val key = arguments.uid

        removeAdIfExists(key)

        val ad = createAd(context, arguments)

        ads[key] = ad

        ad.load(
            arguments.parameters,
            eventDispatcher = RewardedAdEventDispatcherFacade(
                arguments.uid,
                eventDispatcher
            )
        )
    }

    fun showAd(uid: String) {
        val ad = ads[uid] ?: throw Exception("Can not show ad. It has to be loaded first")

        ad.show()
    }

    fun removeAd(uid: String) {
        removeAdIfExists(uid)
    }

    private fun createAd(
        context: Context,
        arguments: RewardedYandexAdArguments,
    ): RewardedYandexAd {
        return RewardedYandexAd(
            context = context,
            adId = arguments.adId,
        )
    }

    private fun removeAdIfExists(key: String) {
        val existingAd = ads[key] ?: return

        existingAd.destroy()
        ads.remove(key)
    }
}

File: ios/Classes/platform_api/yandex_ads_organizer/RewardedYandexAdsOrganizer.swift
//
//  InterstitialYandexAdOrganizer.swift
//  flutter_yandex_ads_plus
//
//  Created by apple on 23.01.2023.
//

import Foundation

class RewardedYandexAdsOrganizer {
    private let eventDispatcher: AdEventDispatcher
    private var ads: [String: RewardedYandexAd] = [:]
    
    init(eventDispatcher: AdEventDispatcher) {
        self.eventDispatcher = eventDispatcher
    }
    
    func createAndLoadAd(args: RewardedAdArguments) {
        let key = args.uid

        removeAdIfExists(uid: key)

        let ad = createAd(args: args)
        ads[key] = ad
        
        ad.load(parameters: args.adParameters)
    }
    
    func showAd(uid: String) throws {
        if let ad = ads[uid] {
            try ad.show()
        } else {
            throw PluginError.runtimeError(
                "Ad can not be shown. It has to be loaded first"
            )
        }
    }
    
    func removeAd(uid: String) {
        if let ad = ads.removeValue(forKey: uid) {
            ad.destroy()
        }
    }
    
    private func createAd(args: RewardedAdArguments)
    -> RewardedYandexAd {
        return RewardedYandexAd(
            adUnitId: args.adId,
            eventDispatcher: RewardedAdEventDispatcherFacade(
                uid: args.uid,
                dispatcher: eventDispatcher
            )
        )
    }
    
    private func removeAdIfExists(uid: String) {
        ads.removeValue(forKey: uid)
    }
}
